### PR TITLE
Define and expose user-defined error handler

### PR DESCRIPTION
**Pull request by flomnes**

### CODE REVIEW

File: src/libzippp.h [hunk 1]
```diff
@@ -538,6 +547,8 @@ namespace libzippp {
         
         void** bufferData;
         libzippp_uint64 bufferLength;
+
+        ErrorHandlerCallback errorHandlingCallback = nullptr;
```
Comment (ctabin) [inline src/libzippp.h:551]: Initialize the variable in the constructor initializer.

File: src/libzippp.h [hunk 2]
```diff
@@ -526,6 +531,10 @@ namespace libzippp {
         inline double getProgressPrecision(void) const { return progressPrecision; }
         void setProgressPrecision(double p) { progressPrecision = p; }
 
+        void setErrorHandlerCallback(const ErrorHandlerCallback& callback) {
```
Comment (ctabin) [inline src/libzippp.h:534]: Add documentation.

File: src/libzippp.h [hunk 3]
```diff
@@ -99,14 +99,19 @@ struct zip_source;
 namespace libzippp {
     class ZipEntry;
     class ZipProgressListener;
-    
+    /**
```
Comment (ctabin) [inline src/libzippp.h:102]: Add an empty line before.

File: src/libzippp.h [hunk 4]
```diff
@@ -99,14 +99,19 @@ struct zip_source;
...
+     * User-defined error-handler.
+     * See https://libzip.org/documentation/zip_error_system_type.html
+     */
+    using ErrorHandlerCallback = std::function<void(int /* error_code_zip */,
+                                                    int /* error_code_system */)>;
```
Comment (ctabin) [inline src/libzippp.h:107]: I would suggest to define variables name here, it is more readable IMHO.

```suggestion
    using ErrorHandlerCallback = std::function<void(int zip_error_code,
                                                    int system_error_code)>;
```
Maybe also add a small description about those two codes.

File: src/libzippp.h [hunk 5]
```diff
@@ -99,14 +99,19 @@ struct zip_source;
...
+
     /**
      * Represents a ZIP archive. This class provides useful methods to handle an archive
      * content. It is simply a wrapper around libzip.
      */
     class LIBZIPPP_API ZipArchive {
     public:
-
```
Comment (ctabin) [inline src/libzippp.h:109]: Please let this empty line.

File: src/libzippp.cpp [hunk 6]
```diff
@@ -310,7 +330,11 @@ int ZipArchive::close(void) {
                         zip_int64_t newLength = bufferLength + increment;
                         sourceBuffer = realloc(sourceBuffer, newLength * sizeof(char));
                         if(sourceBuffer==nullptr) {
+                          if (errorHandlingCallback) {
+                            callErrorHandlingCb(zipHandle, errorHandlingCallback);
```
Comment (ctabin) [inline src/libzippp.cpp:334]: Beware of the indentation.

**Review by ctabin — CHANGES_REQUESTED** [review]
@flomnes Thanks for the PR ! I left a few comments. Can you also rebase, since there is a conflict.

File: src/libzippp.h (hunk 1 above)
Comment (flomnes) [reply src/libzippp.h:551]: done

File: src/libzippp.h (hunk 3 above)
Comment (flomnes) [reply src/libzippp.h:102]: done

File: src/libzippp.h (hunk 5 above)
Comment (flomnes) [reply src/libzippp.h:109]: done

File: src/libzippp.cpp (hunk 6 above)
Comment (flomnes) [reply src/libzippp.cpp:332]: done

File: src/libzippp.h (hunk 2 above)
Comment (flomnes) [reply src/libzippp.h:534]: Done, see README.md

**Comment by flomnes** [comment]
Hi @ctabin, I've applied the requested changes. I've also added a forgotten call to the callback, see [e0cdedc](https://github.com/ctabin/libzippp/pull/156/commits/e0cdedc0fbb850b2a4b3e233811e9fc348b97879).

**Comment by ctabin** [comment]
@flomnes LGTM. Thanks for the amazing work !

**Pull request merged**